### ISSUE TITLE
Add a way to check for duplicate dependencies

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/cargo#7285]
There should be a simple way to check that there aren't any duplicate dependencies in the dependency graph, e.g., a `cargo build --unique-deps` / `cargo run --unique-deps` that errors if two versions of a dependency are anywhere in the dependency graph. 

This would be useful for usage in CI. 

What I currently do, is manually use `cargo tree` to pretty print the whole dependency tree, and manually inspect the dependencies. There should be a better way.

**Comment by est31** [comment]
@RalfJung has used Cargo.lock for doing this but the upcoming format doesn't lend itself as nicely to detection of dupes and it was suggested by to use tooling for this instead: https://github.com/rust-lang/cargo/pull/7070#issuecomment-511444061

It would also be very helpful if you could specify a whitelist of crates that are allowed to be duplicate. Sometimes it requires major effort to deduplicate a dependency. I think servo might only switch to the builtin tool if you can specify a whitelist. 

cc @Eijebong

**Comment by Eijebong** [comment]
Yeah, it's really really hard to go for 0 dupes. (try to have only one version of rand_core if you've got anything depending on rand or one version of winapi if you've got anything depending on mio...).

A whitelist is needed but at least makes people aware of the fact they're duping something. It's useless without a committed lockfile though and that's something to keep in mind.

**Comment by Aaron1011** [comment]
@gnzlbg: You can use `cargo tree -d` to only print duplicates.

**Comment by epage** [comment]
I'd love for us to look into moving some of `cargo-deny`s functionality into cargo (or [cargo-clippy](https://rust-lang.zulipchat.com/#narrow/stream/246057-t-cargo/topic/Cargo.20Lints)) once we have #12235.

**Comment by epage** [comment]
As this is more general ("add a way") and `cargo tree -d` technically fulfills that, I'm going to close this with #13899 tracking the linting side of this.
